### PR TITLE
Rework ChecksummedBytes internals to use a Range instead of a Bytes slice

### PR DESCRIPTION
**Pull request by passaro**
## Description of change

Preliminary refactor to prepare for adding integrity checks on the range itself. No changes in behavior.

## Does this change impact existing behavior?

No.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3/src/checksums.rs [hunk 1]
```diff
@@ -275,26 +316,69 @@ mod tests {
...
+    fn slice(original: Range<usize>, range: impl RangeBounds<usize>) -> ChecksummedBytes {
+        let buffer = Bytes::copy_from_slice(&vec![0; original.len()]);
+        let checksum = crc32c::checksum(&buffer);
+        let bytes = ChecksummedBytes {
+            buffer,
+            range: original,
+            checksum,
+        };
+        bytes.slice(range)
+    }
```
Comment (dannycjones) [inline mountpoint-s3/src/checksums.rs:336]: I'd prefer we just include this setup inline in the test. We can simplify the code a bit too:

```rust
let buffer = Bytes::from_static([0; original.len()]);
let bytes = ChecksummedBytes::from_bytes(buffer);
let slice = bytes.slice(range);
```

Comment (dannycjones) [reply mountpoint-s3/src/checksums.rs:336]: better yet, we could just add (maybe test only) `ChecksummedBytes::from_static`? test setup / 'boilerplate' becomes:

```rust
let bytes = ChecksummedBytes::from_static([0; original.len()]);
let slice = bytes.slice(range);
```

**Review by dannycjones — CHANGES_REQUESTED** [review]
about there, just two minor things

File: mountpoint-s3/src/checksums.rs (hunk 1 above)
Comment (passaro) [reply mountpoint-s3/src/checksums.rs]: Your suggestion is missing the `original` range, but you are definitely right about the `slice` helper function not being clear! 
In fact, I think the issue is that it is both setting up the instance to test and invoking the `slice` method on it. I've fixed that by using the helper function just for the setup and calling `slice` in each test function. I hope that it's clearer now.

Comment (dannycjones) [reply mountpoint-s3/src/checksums.rs]: thanks, yep clearer!

**Review by dannycjones — APPROVED** [review]
lgtm!

**Pull request merged**